File: src/main/java/com/flexpoker/table/command/handlers/StartNewHandForExistingTableCommandHandler.java
package com.flexpoker.table.command.handlers;

import javax.inject.Inject;

import org.springframework.scheduling.annotation.Async;
import org.springframework.stereotype.Component;

import com.flexpoker.framework.command.CommandHandler;
import com.flexpoker.framework.event.EventPublisher;
import com.flexpoker.table.command.commands.StartNewHandForExistingTableCommand;
import com.flexpoker.table.command.factory.TableFactory;
import com.flexpoker.table.command.framework.TableEvent;
import com.flexpoker.table.command.repository.TableEventRepository;
import com.flexpoker.table.command.service.CardService;
import com.flexpoker.table.command.service.HandEvaluatorService;

@Component
public class StartNewHandForExistingTableCommandHandler implements
        CommandHandler<StartNewHandForExistingTableCommand> {

    private final TableFactory tableFactory;

    private final EventPublisher<TableEvent> eventPublisher;

    private final TableEventRepository tableEventRepository;

    private final CardService cardService;

    private final HandEvaluatorService handEvaluatorService;

    @Inject
    public StartNewHandForExistingTableCommandHandler(TableFactory tableFactory,
            EventPublisher<TableEvent> eventPublisher,
            TableEventRepository tableEventRepository, CardService cardService,
            HandEvaluatorService handEvaluatorService) {
        this.tableFactory = tableFactory;
        this.eventPublisher = eventPublisher;
        this.tableEventRepository = tableEventRepository;
        this.cardService = cardService;
        this.handEvaluatorService = handEvaluatorService;
    }

    @Async
    @Override
    public void handle(StartNewHandForExistingTableCommand command) {
        var tableEvents = tableEventRepository.fetchAll(command.getTableId());
        var table = tableFactory.createFrom(tableEvents);

        var shuffledDeckOfCards = cardService.createShuffledDeck();
        var cardsUsedInHand = cardService.createCardsUsedInHand(shuffledDeckOfCards, table.getNumberOfPlayersAtTable());

        var possibleHandRankings = handEvaluatorService.determinePossibleHands(
                cardsUsedInHand.getFlopCards(), cardsUsedInHand.getTurnCard(), cardsUsedInHand.getRiverCard());
        var handEvaluations = handEvaluatorService.determineHandEvaluation(
                cardsUsedInHand.getFlopCards(), cardsUsedInHand.getTurnCard(), cardsUsedInHand.getRiverCard(),
                cardsUsedInHand.getPocketCards(), possibleHandRankings);

        table.startNewHandForExistingTable(command.getSmallBlind(),
                command.getBigBlind(), shuffledDeckOfCards, cardsUsedInHand,
                handEvaluations);
        table.fetchNewEvents().forEach(x -> tableEventRepository.save(x));
        table.fetchNewEvents().forEach(x -> eventPublisher.publish(x));
    }

}

File: src/main/kotlin/com/flexpoker/table/command/handlers/StartNewHandForExistingTableCommandHandler.kt
package com.flexpoker.table.command.handlers

import com.flexpoker.framework.command.CommandHandler
import com.flexpoker.framework.event.EventPublisher
import com.flexpoker.table.command.aggregate.applyEvents
import com.flexpoker.table.command.aggregate.eventproducers.numberOfPlayersAtTable
import com.flexpoker.table.command.aggregate.eventproducers.startNewHandForExistingTable
import com.flexpoker.table.command.commands.StartNewHandForExistingTableCommand
import com.flexpoker.table.command.events.TableEvent
import com.flexpoker.table.command.repository.TableEventRepository
import com.flexpoker.table.command.service.CardService
import com.flexpoker.table.command.service.HandEvaluatorService
import org.springframework.scheduling.annotation.Async
import org.springframework.stereotype.Component
import javax.inject.Inject

@Component
class StartNewHandForExistingTableCommandHandler @Inject constructor(
    private val eventPublisher: EventPublisher<TableEvent>,
    private val tableEventRepository: TableEventRepository,
    private val cardService: CardService,
    private val handEvaluatorService: HandEvaluatorService
) : CommandHandler<StartNewHandForExistingTableCommand> {

    @Async
    override fun handle(command: StartNewHandForExistingTableCommand) {
        val existingEvents = tableEventRepository.fetchAll(command.tableId)
        val state = applyEvents(existingEvents)
        val shuffledDeckOfCards = cardService.createShuffledDeck()
        val cardsUsedInHand = cardService.createCardsUsedInHand(shuffledDeckOfCards, numberOfPlayersAtTable(state))
        val handEvaluations = handEvaluatorService.determineHandEvaluation(
            cardsUsedInHand.flopCards, cardsUsedInHand.turnCard, cardsUsedInHand.riverCard,
            cardsUsedInHand.pocketCards)
        val newEvents = startNewHandForExistingTable(state, command.smallBlind, command.bigBlind,
            shuffledDeckOfCards, cardsUsedInHand, handEvaluations)
        val newlySavedEventsWithVersions = tableEventRepository.setEventVersionsAndSave(existingEvents.size, newEvents)
        newlySavedEventsWithVersions.forEach { eventPublisher.publish(it) }
    }

}